### PR TITLE
feat: Add AppVersionCheck feature

### PR DESCRIPTION
**Pull request by BiancaMihaila**
**Ticket Link:** 
[DCMAW-7426](https://govukverify.atlassian.net/browse/DCMAW-7426)

**Description Of Changes:** 
- create use case for `AppVersionCheck` and implement it within the `AppInfoService.get()`
- update `SplashScreenViewModel.retrieveAppInfo()` to handle the check and logic
- update `NavGraphObjects` tests to use AndroidComposeTestRule instead on PhoneController   
- refactor `UpdateRequiredScreen` -> `UpdateRequiredErrorScreen` and move into the ui.error package - made it part of the `ErrorGraphObject`

**Evidence:**

**_AC1_**

https://github.com/user-attachments/assets/0b79d582-2f5f-4874-819c-6ae70ab1be7c

**_AC4_**

[app_min_v_update_req_cy.webm](https://github.com/user-attachments/assets/7811fe59-fef3-4aad-99c5-aaf5d01504eb)

_**AC2b & AC3**_
This has been build with a temporary dev stack - not using the / appInfo url for the request - the logic is the same and it should have the same behaviour when using the correct url. This is the response received from the endpoint:
```
{
  "apps": {
    "android": {
      "minimumVersion": "0.0.1",
      "releaseFlags": {
        "walletVisibleViaDeepLink": true,
        "walletVisibleIfExists": true,
        "walletVisibleToAll": false
      },
      "available": true,
      "featureFlags" : {
        "appCheckEnabled": false
      }
    },
    "iOS": {
      "minimumVersion": "1.0.0",
      "releaseFlags": {
        "walletVisibleViaDeepLink": true,
        "walletVisibleIfExists": true,
        "walletVisibleToAll": false
      },
      "available": true,
      "featureFlags" : {
        "appCheckEnabled": true
      }
    }
  }
}
```

https://github.com/user-attachments/assets/fc85b39c-b798-42cf-8d3a-efe3fd8e0514

**Definition Of Done Checklist:**

- [x] Unit tests written to at least 80% code coverage (unless explicit approval given from TL that this is not necessary)
- [x] Relevant integration and end-to-end tests are written
- [x] Tests cover each acceptance criteria on the user story
- [ ] Demo completed by running code locally, demoing to a code reviewer & designer
- [x] Sonar coverage gates met 
- [ ] If feature flagged, Feature is enabled in staging and manually tested to ensure integration

Resolves: DCMAW-7426

[DCMAW-7426]: https://govukverify.atlassian.net/browse/DCMAW-7426?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: app/src/androidTest/java/uk/gov/onelogin/login/ui/splash/SplashScreenTest.kt [hunk 1]
```diff
@@ -43,25 +48,43 @@ class SplashScreenTest : TestCase() {
...
+        loadingIndicator = hasContentDescription(
+            resources.getString(R.string.app_splashScreenLoadingContentDescription)
+        )
+        loadingText = hasText(resources.getString(R.string.app_splashScreenLoadingIndicatorText))
     }
 
     @Test
     fun verifySplashScreen() {
         // Given
+        wheneverBlocking { appInfoService.get() }.thenReturn(
```
Comment (alex-bradbury-gds) [inline app/src/androidTest/java/uk/gov/onelogin/login/ui/splash/SplashScreenTest.kt:80]: this already happens in the `@Before` ?

File: app/src/main/java/uk/gov/onelogin/appinfo/appversioncheck/data/AppVersionCheckImpl.kt [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+            }
+        } catch (e: AppInfoUtils.AppError) {
+            Log.e(TAG, e.toString())
+            result = AppInfoServiceState.Unavailable
+        }
+        return result
+    }
+
+    companion object {
+        const val TAG = "AppVersionCheck"
```
Comment (alex-bradbury-gds) [inline app/src/main/java/uk/gov/onelogin/appinfo/appversioncheck/data/AppVersionCheckImpl.kt:38]: this should probably be private

File: app/src/main/java/uk/gov/onelogin/appinfo/service/data/AppInfoServiceImpl.kt [hunk 3]
```diff
@@ -13,15 +14,23 @@ import uk.gov.onelogin.appinfo.source.domain.source.AppInfoRemoteSource
...
             is AppInfoRemoteState.Success -> {
-                val encodedValue = Json.encodeToString<AppInfoData>(remoteResult.value)
-                this.localSource.update(encodedValue)
-                return AppInfoServiceState.RemoteSuccess(remoteResult.value)
+                // Check remote min version compatible
+                when (val compareResult = appVersionCheck.compareVersions(remoteResult.value)) {
+                    is AppInfoServiceState.Successful -> {
+                        // Encode value to save in SharedPrefs and save/ update
+                        val encodedValue = Json.encodeToString<AppInfoData>(remoteResult.value)
+                        localSource.update(encodedValue)
```
Comment (alex-bradbury-gds) [inline app/src/main/java/uk/gov/onelogin/appinfo/service/data/AppInfoServiceImpl.kt:29]: This section in general; I wonder if we want to still save the app info locally even if the app needs updating 🤷  then this bit could look something like:
```
localSource.update(remoteResult.value)
return appVersionCheck.compareVersions(remoteResult.value)
```

File: app/src/main/java/uk/gov/onelogin/appinfo/service/domain/model/AppInfoServiceState.kt [hunk 4]
```diff
@@ -3,8 +3,8 @@ package uk.gov.onelogin.appinfo.service.domain.model
 import uk.gov.onelogin.appinfo.apicall.domain.model.AppInfoData
 
 sealed class AppInfoServiceState {
-    data class RemoteSuccess(val value: AppInfoData) : AppInfoServiceState()
-    data class LocalSuccess(val value: AppInfoData) : AppInfoServiceState()
+    data class Successful(val value: AppInfoData) : AppInfoServiceState()
```
Comment (alex-bradbury-gds) [inline app/src/main/java/uk/gov/onelogin/appinfo/service/domain/model/AppInfoServiceState.kt:6]: nice, think this simplification is good

File: app/src/main/java/uk/gov/onelogin/appinfo/service/data/AppInfoServiceImpl.kt [hunk 5]
```diff
@@ -34,7 +43,9 @@ class AppInfoServiceImpl @Inject constructor(
         fallback: AppInfoServiceState
     ): AppInfoServiceState {
         return if (localSourceState is AppInfoLocalState.Success) {
-            AppInfoServiceState.LocalSuccess(localSourceState.value)
+            // Check local min version compatible
+            appVersionCheck.compareVersions(localSourceState.value)
```
Comment (alex-bradbury-gds) [inline app/src/main/java/uk/gov/onelogin/appinfo/service/data/AppInfoServiceImpl.kt:47]: Can we just return this line? Given its under local success already it _should_ give a success back anyway

File: app/src/main/java/uk/gov/onelogin/login/ui/splash/SplashScreenViewModel.kt [hunk 6]
```diff
@@ -68,14 +76,15 @@ class SplashScreenViewModel @Inject constructor(
         navigator.navigate(LoginRoutes.AnalyticsOptIn)
     }
 
-    fun retrieveAppInfo() {
+    fun retrieveAppInfo(callback: suspend CoroutineScope.() -> Unit) {
```
Comment (alex-bradbury-gds) [inline app/src/main/java/uk/gov/onelogin/login/ui/splash/SplashScreenViewModel.kt:79]: Could rename `callback` here to make it obvious that's it's an 'onSuccess` callback

File: app/src/test/java/uk/gov/onelogin/TestUtils.kt [hunk 7]
```diff
@@ -0,0 +1,50 @@
+package uk.gov.onelogin
+
+import uk.gov.onelogin.appinfo.apicall.domain.model.AppInfoData
+
+object TestUtils {
+    val data = AppInfoData(
```
Comment (alex-bradbury-gds) [inline app/src/test/java/uk/gov/onelogin/TestUtils.kt:6]: again, maybe just a rename here to something like `appInfoData`

File: app/src/androidTest/java/uk/gov/onelogin/TestUtils.kt [hunk 8]
```diff
@@ -2,11 +2,33 @@ package uk.gov.onelogin
 
 import androidx.compose.ui.test.junit4.AndroidComposeTestRule
 import androidx.test.ext.junit.rules.ActivityScenarioRule
+import uk.gov.onelogin.appinfo.apicall.domain.model.AppInfoData
 
 object TestUtils {
-    fun AndroidComposeTestRule<ActivityScenarioRule<MainActivity>, MainActivity>.setNavInitialPoint(
+    val data = AppInfoData(
```
[the remark this answers is not in the document]
Comment (BiancaMihaila) [reply app/src/androidTest/java/uk/gov/onelogin/TestUtils.kt]: makes sense, done now, thanks

File: app/src/androidTest/java/uk/gov/onelogin/login/ui/splash/SplashScreenTest.kt (hunk 1 above)
Comment (BiancaMihaila) [reply app/src/androidTest/java/uk/gov/onelogin/login/ui/splash/SplashScreenTest.kt]: Sorry, forgot to remove this after I added it to the `@Before`, done

**Comment by co-eoin** [comment]
QA Reviewed & Approved ✅

**Review by alex-bradbury-gds — APPROVED** [review]
Mega stuff 💪

**Pull request merged**